### PR TITLE
perf: optimize msm gpu memory usage

### PR DESCRIPTION
**Pull request by GideokKim**
# Description

This PR is a performance improvement that allows the base to be recycled after initial allocation without reallocating when using MSM GPU in kzg commitment scheme.

### CODE REVIEW

File: tachyon/crypto/commitments/kzg/kzg.h [hunk 1]
```diff
@@ -62,7 +63,7 @@ class KZG {
     CHECK_EQ(g1_powers_of_tau_.size(), g1_powers_of_tau_lagrange_.size());
     CHECK_LE(g1_powers_of_tau_.size(), kMaxDegree + 1);
 #if TACHYON_CUDA
-    SetupForGpu();
+    SetupForGpu(g1_powers_of_tau_.size());
```
Comment (chokobole) [inline tachyon/crypto/commitments/kzg/kzg.h:66]: Since `g1_powers_of_tau_` is a member variable, you don't need to pass this as an argument.

File: tachyon/device/gpu/gpu_memory.h [hunk 2]
```diff
@@ -173,14 +173,14 @@ class GpuMemory {
...
 
-  T* operator->() { return get(); }
-  const T* operator->() const { return get(); }
+  T* operator->() { return data(); }
+  const T* operator->() const { return data(); }
 
-  T* get() { return ptr_; }
-  const T* get() const { return ptr_; }
+  T* data() { return ptr_; }
+  const T* data() const { return ptr_; }
```
Comment (chokobole) [inline tachyon/device/gpu/gpu_memory.h:183]: Why did you change this name to `data()`?

File: tachyon/device/gpu/gpu_memory.h [hunk 3]
```diff
@@ -173,14 +173,14 @@ class GpuMemory {
...
+  T& operator*() { return *data(); }
+  const T& operator*() const { return *data(); }
 
-  T* operator->() { return get(); }
-  const T* operator->() const { return get(); }
+  T* operator->() { return data(); }
+  const T* operator->() const { return data(); }
 
-  T* get() { return ptr_; }
-  const T* get() const { return ptr_; }
```
Comment (batzor) [inline tachyon/device/gpu/gpu_memory.h:183]: Since you only rename `get()` to `data()` I think you can reword the commit to 
```
refac(device): rename `GpuMemory::get()` to `GpuMemory::data()`
```

File: tachyon/device/gpu/gpu_memory.h (hunk 2 above)
Comment (GideokKim) [reply tachyon/device/gpu/gpu_memory.h:183]: The `std::vector` does not have a `get` member function, so I initially attempted to have `DoMSM`'s bases call a member function with the same name. Upon further review, I realized this approach was not suitable.

A cleaner solution would be to change `bases` argument used by `DoMSM` to `absl::Span<const G1Point>` within the `CommitLagrange()` and `Commit()` functions. Consequently, `DoMSM` should receive `absl::Span<const G1Point>` as an argument.

**Comment by TomTaehoonKim** [comment]
@GideokKim "feat(math): add base copy option change" What does this commit message mean? I really can't relate well with the diff.

File: tachyon/crypto/commitments/kzg/kzg.h [hunk 4]
```diff
@@ -89,8 +90,14 @@ class KZG {
     CHECK_EQ(error, gpuSuccess);
     stream_ = device::gpu::CreateStream();
 
+    size_t bases_size = g1_powers_of_tau_.size();
     msm_gpu_.reset(
         new math::VariableBaseMSMGpu<G1Point>(mem_pool_.get(), stream_.get()));
+    d_g1_powers_of_tau_ = device::gpu::GpuMemory<G1Point>::MallocFromPoolAsync(
```
Comment (TomTaehoonKim) [inline tachyon/crypto/commitments/kzg/kzg.h:96]: What does the "d" prefix stand for?

**Comment by TomTaehoonKim** [comment]
f26c389: refac(device): refac(device): -> refac(device):

File: tachyon/zk/plonk/examples/fibonacci/BUILD.bazel [hunk 5]
```diff
@@ -73,7 +73,6 @@ tachyon_cc_library(
 
 tachyon_cc_unittest(
     name = "fibonacci1_circuit_test",
-    testonly = True,
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/examples/fibonacci/BUILD.bazel:76]: Why is it redundant?

File: tachyon/device/gpu/gpu_memory.h (hunk 3 above)
Comment (batzor) [reply tachyon/device/gpu/gpu_memory.h:183]: This is still not fixed btw.

File: tachyon/crypto/commitments/kzg/kzg.h (hunk 4 above)
Comment (ashjeong) [reply tachyon/crypto/commitments/kzg/kzg.h:96]: I think it means "device"since we have `device::gpu::GpuMemory<G1Point> d_g1_powers_of_tau_;`

File: tachyon/device/gpu/gpu_memory.h (hunk 3 above)
Comment (ashjeong) [reply tachyon/device/gpu/gpu_memory.h:183]: I think it's fixed but "refac(device):" is written twice. Please fix!

File: tachyon/zk/plonk/examples/fibonacci/BUILD.bazel (hunk 5 above)
Comment (ashjeong) [reply tachyon/zk/plonk/examples/fibonacci/BUILD.bazel:76]: I think it's because it's a `tachyon_cc_unittest` which ensures it's a test lol.

File: tachyon/device/gpu/gpu_memory.h (hunk 2 above)
Comment (GideokKim) [reply tachyon/device/gpu/gpu_memory.h:183]: I realized that my previous method was not a good idea either. In the code below, I confirmed that the argument is passed as `std::vector` even if `msm_gpu_` exists normally.

- `tachyon/zk/base/commitments/gwc_extension.h`
- `tachyon/zk/base/commitments/shplonk_extension.h`

I believe there are two options:

1. Maintain the current status.
2. When using cuda build, modify the code to return `d_g1_powers_of_tau_` or `d_g1_powers_of_tau_lagrange_` allocated to device memory through the `g1_powers_of_tau()` and `g1_powers_of_tau_lagrange()` member functions.

Which plan is best?

File: tachyon/crypto/commitments/kzg/kzg.h (hunk 4 above)
Comment (GideokKim) [reply tachyon/crypto/commitments/kzg/kzg.h:96]: @ashjeong  correct!

@TomTaehoonKim When naming to a variable for memory allocated to device(GPU), prefix its name with `d_`. Below is an example of NVIDIA naming in ICICLE:

- [ICICLE example](https://github.com/ingonyama-zk/icicle/blob/main/icicle/src/msm/msm.cu#L413)
- [NVIDIA CUB example](https://github.com/NVIDIA/cub/blob/main/examples/device/example_device_radix_sort.cu#L194-L195)

File: tachyon/zk/plonk/examples/fibonacci/BUILD.bazel (hunk 5 above)
Comment (GideokKim) [reply tachyon/zk/plonk/examples/fibonacci/BUILD.bazel:76]: If you check the [Bazel documentation](https://bazel.build/reference/be/common-definitions#typical-attributes), you will find the following:

Tests (`*_test` rules) and test suites (`test_suite` rules) are `testonly` by default.

Additional reference:

```python
def tachyon_cc_unittest(
        name,
        size = "small",
        **kwargs):
    tachyon_cc_test(
        name = name,
        size = size,
        **kwargs
    )
    
def tachyon_cc_test(
        name,
        copts = [],
        defines = [],
        local_defines = [],
        linkopts = [],
        linkstatic = True,
        deps = [],
        safe_code = True,
        force_exceptions = False,
        force_rtti = False,
        **kwargs):
    cc_test(
        name = name,
        copts = copts + tachyon_cxxopts(safe_code = safe_code, force_exceptions = force_exceptions, force_rtti = force_rtti),
        defines = defines + tachyon_defines(),
        local_defines = local_defines + tachyon_local_defines(),
        linkopts = linkopts + tachyon_linkopts(),
        linkstatic = linkstatic,
        deps = deps + ["@com_google_googletest//:gtest_main"],
        **kwargs
    )
```

**Review by batzor — APPROVED** [review]
LGTM

**Review by ashjeong — APPROVED** [review]
LGTM

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

File: tachyon/math/elliptic_curves/msm/algorithms/icicle/icicle_msm_bls12_381_g2.cc [hunk 6]
```diff
@@ -55,8 +65,7 @@ bool IcicleMSM<bls12_381::G2AffinePoint>::Run(
     ::bls12_381::g2_projective_t ret;
     gpuError_t error = tachyon_bls12_381_g2_msm_cuda(
         reinterpret_cast<const ::bls12_381::scalar_t*>(&cpu_scalars[start_idx]),
-        reinterpret_cast<const ::bls12_381::g2_affine_t*>(
-            &bases[start_idx]),
+        reinterpret_cast<const ::bls12_381::g2_affine_t*>(&bases[start_idx]),
```
Comment (dongchangYoo) [inline tachyon/math/elliptic_curves/msm/algorithms/icicle/icicle_msm_bls12_381_g2.cc:68]: It looks like there is a diff even though there are no changes. I need to check the linter.

**Review by dongchangYoo — CHANGES_REQUESTED** [review]

File: tachyon/zk/base/commitments/BUILD.bazel [hunk 7]
```diff
@@ -8,6 +8,7 @@ tachyon_cc_library(
     deps = [
         ":univariate_polynomial_commitment_scheme_extension",
         "//tachyon/crypto/commitments/kzg:gwc",
+        "//tachyon/device/gpu:gpu_memory",
```
Comment (chokobole) [inline tachyon/zk/base/commitments/BUILD.bazel:11]: I think this dependency seems to be redundant. See the reason below.

File: tachyon/zk/base/commitments/gwc_extension.h [hunk 8]
```diff
@@ -16,6 +16,10 @@
 #include "tachyon/crypto/commitments/kzg/gwc.h"
 #include "tachyon/zk/base/commitments/univariate_polynomial_commitment_scheme_extension.h"
 
+#if TACHYON_CUDA
+#include "tachyon/device/gpu/gpu_memory.h"
+#endif
```
Comment (chokobole) [inline tachyon/zk/base/commitments/gwc_extension.h:21]: I think this is redundant. Since `gwc_.kzg().d_g1_powers_of_tau()` returns `device::gpu::GpuMemory<G1Point>`, it's safe to assume that the necessary header is already included with the implementation of kzg.

File: tachyon/c/zk/plonk/halo2/BUILD.bazel [hunk 9]
```diff
@@ -176,6 +176,7 @@ tachyon_cc_library(
         "//tachyon/base:logging",
         "//tachyon/c/base:type_traits_forward",
         "//tachyon/c/math/elliptic_curves:point_traits_forward",
+        "//tachyon/device/gpu:gpu_memory",
```
Comment (chokobole) [inline tachyon/c/zk/plonk/halo2/BUILD.bazel:179]:
```suggestion
```

File: tachyon/c/zk/plonk/halo2/kzg_family_prover_impl.h [hunk 10]
```diff
@@ -13,6 +13,10 @@
 #include "tachyon/c/zk/plonk/halo2/prover_impl_base.h"
 #include "tachyon/math/elliptic_curves/msm/variable_base_msm.h"
 
+#if TACHYON_CUDA
+#include "tachyon/device/gpu/gpu_memory.h"
+#endif
```
Comment (chokobole) [inline tachyon/c/zk/plonk/halo2/kzg_family_prover_impl.h:18]: This is redundant since this file isn't using any symbol or type in the header.

File: tachyon/c/zk/plonk/halo2/kzg_family_prover_impl.h [hunk 11]
```diff
@@ -32,16 +36,25 @@ class KZGFamilyProverImpl : public ProverImplBase<PS> {
...
+    return DoMSM(this->pcs_.GetDeviceG1PowersOfTau(), scalars);
+#else
     return DoMSM(this->pcs_.GetG1PowersOfTau(), scalars);
+#endif
   }
 
   CProjectivePoint* CommitLagrangeRaw(
       const std::vector<ScalarField>& scalars) const {
+#if TACHYON_CUDA
+    return DoMSM(this->pcs_.GetDeviceG1PowersOfTauLagrange(), scalars);
```
Comment (chokobole) [inline tachyon/c/zk/plonk/halo2/kzg_family_prover_impl.h:49]: Please check if gpu is configured to be run like how we do in KZG code. 

```suggestion
    if (this->pcs_.UsesGPU()) {
      return DoMSM(this->pcs_.GetDeviceG1PowersOfTauLagrange(), scalars);
    }  
```

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Comment by chokobole** [comment]
Please remove scope from [feat(crypto/zk): add gpu memory accessor functions](https://github.com/kroma-network/tachyon/pull/511/commits/5d341d5b50cc80a40a5dd86972a87cc2ef459f38)

**Comment by chokobole** [comment]
Please change **c** to **c/zk** in [feat(c): modify commit functions with gpu support](https://github.com/kroma-network/tachyon/pull/511/commits/876acc58a53ba66650a57be4122cacac280d962c)

File: tachyon/crypto/commitments/kzg/kzg.h [hunk 12]
```diff
@@ -75,6 +75,18 @@ class KZG {
...
+#if TACHYON_CUDA
+  const device::gpu::GpuMemory<G1Point>& d_g1_powers_of_tau() const {
+    return d_g1_powers_of_tau_;
+  }
+
+  const device::gpu::GpuMemory<G1Point>& d_g1_powers_of_tau_lagrange() const {
+    return d_g1_powers_of_tau_lagrange_;
+  }
+
+  bool UsesGPU() const { return msm_gpu_ ? true : false; }
```
Comment (chokobole) [inline tachyon/crypto/commitments/kzg/kzg.h:87]:
```suggestion
  bool UsesGPU() const { return msm_gpu_; }
```

**Review by chokobole — APPROVED** [review]
LGTM

File: tachyon/crypto/commitments/kzg/kzg.h [hunk 13]
```diff
@@ -75,6 +75,18 @@ class KZG {
...
+#if TACHYON_CUDA
+  const device::gpu::GpuMemory<G1Point>& d_g1_powers_of_tau() const {
+    return d_g1_powers_of_tau_;
+  }
+
+  const device::gpu::GpuMemory<G1Point>& d_g1_powers_of_tau_lagrange() const {
+    return d_g1_powers_of_tau_lagrange_;
+  }
+
+  bool UsesGPU() const { return static_cast<bool>(msm_gpu_); }
```
Comment (batzor) [inline tachyon/crypto/commitments/kzg/kzg.h:87]:
```suggestion
  bool UsesGPU() const { return msm_gpu_ != nullptr }
```

File: tachyon/crypto/commitments/kzg/kzg.h [hunk 14]
```diff
@@ -204,24 +204,44 @@ class KZG {
 
   template <typename ScalarContainer>
   [[nodiscard]] bool Commit(const ScalarContainer& v, Commitment* out) const {
+#if TACHYON_CUDA
+    if (msm_gpu_) {
+      return DoMSM(d_g1_powers_of_tau_, v, out);
+    }
+#endif
     return DoMSM(g1_powers_of_tau_, v, out);
```
Comment (batzor) [inline tachyon/crypto/commitments/kzg/kzg.h:224]:
```suggestion
#if TACHYON_CUDA
    if (msm_gpu_) {
      return DoMSM(d_g1_powers_of_tau_, v, out);
    }
#else
    return DoMSM(g1_powers_of_tau_, v, out);
#endif    
```
Please change other parts as well.

**Review by batzor — APPROVED** [review]
LGTM

File: tachyon/crypto/commitments/kzg/kzg.h (hunk 13 above)
Comment (chokobole) [reply tachyon/crypto/commitments/kzg/kzg.h:87]: I think it's find since this is a typical way to check whether the pointer of `unique_ptr<T>` isn't null.

File: tachyon/crypto/commitments/kzg/kzg.h (hunk 14 above)
Comment (chokobole) [reply tachyon/crypto/commitments/kzg/kzg.h:224]: Oh this is intended, because MSM Gpu can't be disabled at any time, at this point, we want to compute it with CPU version.
